libraries/CryptoInterface/src/CryptoInterface.h

```diff
@@ -0,0 +1,806 @@
+
+
+    // #################### MD5+SHA-1 ####################
+    
+    /**
+     * Create a MD5+SHA-1 hash of the data. The result will be MD5SHA1_NATURAL_LENGTH bytes long and stored in resultArray.
+     * Uses the BearSSL cryptographic library.
+     * 
+     * MD5+SHA-1 is the concatenation of MD5 and SHA-1 computed over the same input; in the implementation, the internal data buffer is shared, 
+     * thus making it more memory-efficient than separate MD5 and SHA-1. It can be useful in implementing SSL 3.0, TLS 1.0 and TLS 1.1.
+     * 
+     * @param data The data array from which to create the hash.
+     * @param dataLength The length of the data array in bytes.
+     * @param resultArray The array wherein to store the resulting hash. MUST be be able to contain MD5SHA1_NATURAL_LENGTH bytes or more.
+     * 
+     * @return A pointer to resultArray.
+     */
+    void *md5sha1Hash(const void *data, const size_t dataLength, void *resultArray);
+    
+    /**
+     * Create a MD5+SHA-1 hash of the data. The result will be MD5SHA1_NATURAL_LENGTH bytes long and returned as a String in HEX format.
+     * Uses the BearSSL cryptographic library.
+     * 
+     * MD5+SHA-1 is the concatenation of MD5 and SHA-1 computed over the same input; in the implementation, the internal data buffer is shared, 
+     * thus making it more memory-efficient than separate MD5 and SHA-1. It can be useful in implementing SSL 3.0, TLS 1.0 and TLS 1.1.
+     * 
+     * @param message The string from which to create the hash.
+     * 
+     * @return A String with the generated hash in HEX format.
+     */
+    String md5sha1Hash(const String &message);
+
+
+    // #################### HKDF ####################
+
+    /**
+     * KDFs (key derivation functions) are functions that takes a variable length input, and provide a variable length output, meant to be used to derive subkeys from a master key.
+     * HKDF is a KDF defined by RFC 5869. It is based on HMAC. The provided implementation uses SHA-256 as the underlying hash function.
+     * 
+     * This function initializes the HKDF implementation with the input data to use for HKDF processing.
+     * Uses the BearSSL cryptographic library.
+     * 
+     * Must be called at least once before hkdfProduce() can be used.
+     * 
+     * @param keyMaterial An array containing the key material to use when deriving subkeys. Typically this would be the master key.
+     * @param keyMaterialLength The length of keyMaterial in bytes.
+     * @param salt An array containing the salt to use when ingesting key material. Salt is non-secret and can be empty.
+     *             Its role is normally to bind the input to a conventional identifier that qualify it within the used protocol or application.
+     * @param saltLength The length of the salt array, in bytes.
+     */
+    void hkdfInit(const void *keyMaterial, const size_t keyMaterialLength, const void *salt = nullptr, const size_t saltLength = 0);
```

Comment:
There's obviously some hidden state (since there are separate init and produce calls.  I'd also think this is an object-ification opportunity.